src/notifications/slices/subscriptions.ts

```diff
@@ -0,0 +1,208 @@
+  },
+});
+
+export const subscriptionSelectors = subscriptionsAdapter.getSelectors<State>(
+  (state) => state.notificationSubscriptions
+);
+
+export const fetchSubscriptionById = createAsyncThunk(
+  'notificationSubscriptions/fetchById',
+  async (subscriptionId: number) => {
+    const response = await retrieveNotificationSubscription(subscriptionId);
+    if (response.status === 'success') {
+      return response.data;
+    } else {
+      throw response.errors;
+    }
+  }
+);
+
+export const fetchSubscriptions = createAsyncThunk('notificationSubscriptions/fetchList', async () => {
+  const response = await listNotificationSubscriptions();
+  if (response.status === 'success') {
+    return response.data.results;
+  } else {
+    throw response.errors;
+  }
+});
+
+const transformDeviceSubscription = (subscription: PushSubscription): IDeviceSubscription => {
+  const subscriptionJson = subscription.toJSON();
+  const deviceSubscription = {
+    endpoint: subscriptionJson.endpoint,
+    auth: subscriptionJson?.keys?.auth,
+    p256dh: subscriptionJson?.keys?.p256dh,
+  } as IDeviceSubscription;
+  return deviceSubscription;
+};
+
+export const registerDeviceForPushNotifications = createAsyncThunk(
+  'notificationSubscriptions/registerDevice',
+  async (_, { dispatch }) => {
+    const { subscription, message } = await registerPushManager();
+    if (subscription) {
+      const deviceSubscription = transformDeviceSubscription(subscription);
+      dispatch(subscriptionUpdated({ subscription: deviceSubscription, message }));
+      const response = await createNotificationSubscription(deviceSubscription);
+      if (response.status === 'success') {
+        return response.data;
+      } else {
+        throw response.errors;
+      }
+    } else {
+      dispatch(subscriptionUpdated({ subscription: null, message }));
+      throw message;
+    }
+  }
+);
+
+export const unregisterDeviceFromPushNotfications = createAsyncThunk(
+  'notificationSubscriptions/unregisterDevice',
+  async (_, { getState }) => {
+    const state = getState() as State;
+    const { deviceSubscription } = state.notificationSubscriptions;
+    if (!deviceSubscription) {
+      return;
+    }
+    const subscription = subscriptionSelectors
+      .selectAll(state)
+      .find((sub) => sub.endpoint === deviceSubscription.endpoint);
+    if (!subscription) {
+      return;
+    }
+    const response = await destroyNotificationSubscription(subscription.id);
+    if (response.status !== 'success') {
+      throw response.errors;
+    }
+    const removed = await removeNotificationSubscription();
+    if (removed) {
+      subscriptionUpdated({ subscription: null, message: null });
+    }
+    return subscription.id;
+  }
+);
+
+export const toggleDevicePermission = createAsyncThunk('notificationSubscriptions/toggleDevicePermission', async () => {
+  const allowNotifications = await getNotificationPermission();
+  return allowNotifications;
+});
```

Comment:
It may be a little strange, and I'm not 100% sure since I wrote some of this more than a year ago.
I think it works by asking the user for permission when you try to look at permissions from the code. So here we are asking the browser with code for what the user has selected. If the user has not selected yet, the browser will prompt the user directly.